pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -210,7 +214,7 @@ func (e *EndpointController) addPod(obj interface{}) {
 		return
 	}
 	for key := range services {
-		e.queue.Add(key)
+		e.queue.AddAfter(key, e.endpointUpdatesBatchPeriod)
```

Comment:
Is this the standard way to implement batching? if there is only one endpoint change (probably the most common case), this will delay the endpoint programming by batch period. 

Is it possible to implement batching on the dequeue end? For example:
```
key, ok := e.queue.Get()
if !BatchPeriodAfterLastSync(key) {
   e.queue.Add(key)
   return
}
sync(key)

```

probably need to maintain a map between key and lastSyncTimestamp.

Reply:
> Is this the standard way to implement batching?

It's probably not. But it's certainly the easiest way - it's two lines of code and works.

Regarding your idea - we've been thinking about that and my personal suggestion was that I'm not convinced whether it makes sense to complicate this code (though I don't have that strong opinion).

However, let me throw a bit more data points here:
- I strongly believe we don't need any batching for clusters with less than ~1000 nodes - it should just work once we do things that all under work (EndpointSlice & serialize object once)
- For larger clusters, I also think that the batching period I have on my mind is probably 500ms 
- In clusters with 1000+ nodes, I would generally expect to have non-negligible amount of services, so programming iptables or programming NEGs or sth like that should eat that small 
batching period overhead
- in-cluster network programming latency SLO it's pretty clear that the threshold for SLO will be at least 15s (because that's what we observe even in 100-node clusters in scalability tests)

So this PR is exposing an ability for the operator to tune, with close to zero engineering work and no risk of bugs. The approach you're suggesting doesn't have this properties.

Reply:
This change is the simplest possible way to implement the batching. I agree that there is a number of better ways to find a correct delay (e.g. as you mentioned we can pass the first pod update without delay and delay only next ones, we can use different batch period depending on the factors like observed pod update rate, current size of endpoint etc), but I think this is a good starting point I think.

Please note that we are not changing any default behavior, the change is purely opt-in -- user needs to explicitly set the flag to enable this and can tune the values that makes sense for the given use case.

I think that for large clusters, the reasonable values are around 0.5 seconds as wojtek-t mentioned. While I was testing this with 1s value, I was observing significant (5-10x) reduce in number of endpoints updates generated in scalability tests. Given that endpoints are usually large and observed by all nodes, they generate most of the traffic (in terms of bandwidth and cpu) sent from master.

I think it can be fair tradeoff to increase network programming latency by e.g. 0.5s (out of mentioned 15s) to significantly reduce overall traffic generated on master.